cadquery/occ_impl/exporters/dxf.py

```diff
@@ -1,154 +1,328 @@
+        dxf.document.saveas("rectangle.dxf")
 
-    dxf_spline = ezdxf.math.BSpline(poles, order, knots, weights)
+    .. code-block:: python
+        :caption: Multilayer DXF document
 
-    msp.add_spline().apply_construction_tool(dxf_spline)
+        rectangle = cq.Workplane().rect(10, 20)
+        circle = cq.Workplane().circle(3)
 
+        dxf = DxfDocument()
+        dxf = (
+            dxf.add_layer("layer_1", color=2)
+            .add_layer("layer_2", color=3)
+            .add_shape(rectangle, "layer_1")
+            .add_shape(circle, "layer_2")
+        )
+        dxf.document.saveas("rectangle-with-hole.dxf")
+    """
 
-DXF_CONVERTERS = {
-    "LINE": _dxf_line,
-    "CIRCLE": _dxf_circle,
-    "ELLIPSE": _dxf_ellipse,
-    "BSPLINE": _dxf_spline,
-}
+    CURVE_TOLERANCE = 1e-9
+
+    def __init__(
+        self,
+        dxfversion: str = "AC1027",
+        setup: Union[bool, List[str]] = False,
+        doc_units: int = units.MM,
+        *,
+        metadata: Union[Dict[str, str], None] = None,
+        approx: Optional[ApproxOptions] = None,
+        tolerance: float = 1e-3,
+    ) -> None:
+        """Initialize DXF document.
+
+        :param dxfversion: :attr:`DXF version specifier <ezdxf-stable:ezdxf.document.Drawing.dxfversion>`
+            as string, default is "AC1027" respectively "R2013"
+        :param setup: setup default styles, ``False`` for no setup, ``True`` to set up
+            everything or a list of topics as strings, e.g. ``["linetypes", "styles"]``
+            refer to :func:`ezdxf-stable:ezdxf.new`.
+        :param doc_units: ezdxf document/modelspace :doc:`units <ezdxf-stable:concepts/units>`
+        :param metadata: document :ref:`metadata <ezdxf-stable:ezdxf_metadata>` a dictionary of name value pairs
+        :param approx: Approximation strategy for converting :class:`cadquery.Workplane` objects to DXF entities:
+
+            ``None``
+                no approximation applied
+            ``"spline"``
+                all splines approximated as cubic splines
+            ``"arc"``
+                all curves approximated as arcs and straight segments
+
+        :param tolerance: Approximation tolerance for converting :class:`cadquery.Workplane` objects to DXF entities.
+        """
+        if metadata is None:
+            metadata = {}
+
+        self._DISPATCH_MAP = {
+            "LINE": self._dxf_line,
+            "CIRCLE": self._dxf_circle,
+            "ELLIPSE": self._dxf_ellipse,
+        }
+
+        self.approx = approx
+        self.tolerance = tolerance
+
+        self.document = ezdxf.new(dxfversion=dxfversion, setup=setup, units=doc_units)  # type: ignore[attr-defined]
+        self.msp = self.document.modelspace()
+
+        doc_metadata = self.document.ezdxf_metadata()
+        for key, value in metadata.items():
+            doc_metadata[key] = value
+
+    def add_layer(
+        self, name: str, *, color: int = 1, linetype: str = "CONTINUOUS"
```

Comment:
Could you type linetype using Literal[...]?

Reply:
The value of `linetype` could be any string the user chooses.

Ezdxf creates several [standard linetypes](https://ezdxf.readthedocs.io/en/stable/concepts/linetypes.html) which are supported by all DXF versions, but it is also possible for the user to [create custom line types](https://ezdxf.readthedocs.io/en/stable/tutorials/linetypes.html).

Via the `__init__` `setup=True` parameter a user may choose not to set up standard linetypes.

From https://ezdxf.readthedocs.io/en/stable/concepts/linetypes.html :

> CONTINUOUS is the default linetype for layers with an unspecified linetype.

Reply:
Clear, thanks!